url4c27590fbe_Hoten_Java_Delaunay_tgz-pJ8-com_hoten_delaunay_examples_TestDriverJ8/src/com/hoten/delaunay/voronoi/nodename/as3delaunay/Site.java

```diff
@@ -0,0 +1,371 @@
+        while (i < n && (_edges.get(i).get_visible() == false)) {
+            ++i;
+        }
+
+        if (i == n) {
+            // no edges visible
+            return new ArrayList();
+        }
+        edge = _edges.get(i);
+        LR orientation = _edgeOrientations.get(i);
+        points.add(edge.get_clippedEnds().get(orientation));
+        points.add(edge.get_clippedEnds().get((LR.other(orientation))));
+
+        for (int j = i + 1; j < n; ++j) {
+            edge = _edges.get(j);
+            if (edge.get_visible() == false) {
+                continue;
+            }
+            connect(points, j, bounds, false);
+        }
+        // close up the polygon by adding another corner point of the bounds if needed:
+        connect(points, i, bounds, true);
+
+        return points;
+    }
+
+    private void connect(ArrayList<Point> points, int j, Rectangle bounds, boolean closingUp) {
+        Point rightPoint = points.get(points.size() - 1);
+        Edge newEdge = _edges.get(j);
+        LR newOrientation = _edgeOrientations.get(j);
+        // the point that  must be connected to rightPoint:
+        Point newPoint = newEdge.get_clippedEnds().get(newOrientation);
+        if (!closeEnough(rightPoint, newPoint)) {
+            // The points do not coincide, so they must have been clipped at the bounds;
+            // see if they are on the same border of the bounds:
+            if (rightPoint.x != newPoint.x
+                    && rightPoint.y != newPoint.y) {
+                // They are on different borders of the bounds;
+                // insert one or two corners of bounds as needed to hook them up:
+                // (NOTE this will not be correct if the region should take up more than
+                // half of the bounds rect, for then we will have gone the wrong way
+                // around the bounds and included the smaller part rather than the larger)
+                int rightCheck = BoundsCheck.check(rightPoint, bounds);
+                int newCheck = BoundsCheck.check(newPoint, bounds);
+                double px, py;
+                if ((rightCheck & BoundsCheck.RIGHT) != 0) {
+                    px = bounds.right;
+                    if ((newCheck & BoundsCheck.BOTTOM) != 0) {
+                        py = bounds.bottom;
+                        points.add(new Point(px, py));
+                    } else if ((newCheck & BoundsCheck.TOP) != 0) {
+                        py = bounds.top;
+                        points.add(new Point(px, py));
+                    } else if ((newCheck & BoundsCheck.LEFT) != 0) {
+                        if (rightPoint.y - bounds.y + newPoint.y - bounds.y < bounds.height) {
+                            py = bounds.top;
+                        } else {
+                            py = bounds.bottom;
+                        }
+                        points.add(new Point(px, py));
+                        points.add(new Point(bounds.left, py));
+                    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 243:257, 260:274, 277:291, 294:308.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.